Define in function or method failed to parse

Hi, seems somewhere between requirement`^5.0` and `^5.0@dev` (used in phpDocumentor) (now with commit 7666c4bc) a define declaration into function or method failed to parse.

It seems that cdda55ca introduce this problem, as I don’t get it with the previous 07ce695ce 

Steps to reproduce : 

composer.json
```json
{
    "require": {
        "phpdocumentor/reflection": "5.x-dev#cdda55ca794780f1fe119e16b9f8ff3245d4a5ac"
    }
}
```

test.php
```php
<?php
function test() {
	define('TEST', true);
}
```

demo.php
```php
<?php
include_once 'vendor/autoload.php';
$projectFactory = \phpDocumentor\Reflection\Php\ProjectFactory::createInstance();
$projectFiles = [new \phpDocumentor\Reflection\File\LocalFile('test.php')];
$project = $projectFactory->create('My Project', $projectFiles);
var_dump($project);
```

Run.
```sh
php demo.php
```

That gives:
```
Fatal error: Uncaught AssertionError: assert($file instanceof FileElement) in vendor/phpdocumentor/reflection/src/phpDocumentor/Reflection/Php/Factory/Define.php:103
```

Hope it helps.

--- a/src/phpDocumentor/Reflection/Php/Factory/ContextStack.php
+++ b/src/phpDocumentor/Reflection/Php/Factory/ContextStack.php
@@ -10,6 +10,7 @@ use phpDocumentor\Reflection\Php\File as FileElement;
 use phpDocumentor\Reflection\Php\Project;
 use phpDocumentor\Reflection\Types\Context as TypeContext;
 
+use function array_reverse;
 use function end;
 
 final class ContextStack
@@ -72,5 +73,27 @@ final class ContextStack
         }
 
         return $element;
+    }
+
+    /**
+     * Returns the first element of type.
+     *
+     * Will reverse search the stack for an element matching $type. Will return null when the element type is not
+     * in the current stack.
+     *
+     * @param class-string $type
+     *
+     * @return Element|FileElement|null
+     */
+    public function search(string $type)
+    {
+        $reverseElements = array_reverse($this->elements);
+        foreach ($reverseElements as $element) {
+            if ($element instanceof $type) {
+                return $element;
+            }
+        }
+
+        return null;
     }
 }

--- a/src/phpDocumentor/Reflection/Php/Factory/Define.php
+++ b/src/phpDocumentor/Reflection/Php/Factory/Define.php
@@ -99,7 +99,7 @@ final class Define extends AbstractFactory
             return;
         }
 
-        $file = $context->peek();
+        $file = $context->search(FileElement::class);
         assert($file instanceof FileElement);
 
         $constant = new ConstantElement(

--- a/tests/integration/data/Luigi/constants.php
+++ b/tests/integration/data/Luigi/constants.php
@@ -5,3 +5,7 @@ namespace Luigi;
 const OVEN_TEMPERATURE = 9001;
 define('\\Luigi\\MAX_OVEN_TEMPERATURE', 9002);
 define('OUTSIDE_OVEN_TEMPERATURE', 9002);
+
+function in_function_define(){
+    define('IN_FUNCTION_OVEN_TEMPERATURE', 9003);
+}

--- a/tests/unit/phpDocumentor/Reflection/Php/Factory/ContextStackTest.php
+++ b/tests/unit/phpDocumentor/Reflection/Php/Factory/ContextStackTest.php
@@ -7,6 +7,7 @@ namespace phpDocumentor\Reflection\Php\Factory;
 use OutOfBoundsException;
 use phpDocumentor\Reflection\Fqsen;
 use phpDocumentor\Reflection\Php\Class_ as ClassElement;
+use phpDocumentor\Reflection\Php\Method;
 use phpDocumentor\Reflection\Php\Project;
 use phpDocumentor\Reflection\Types\Context;
 use PHPUnit\Framework\TestCase as PHPUnitTestCase;
@@ -88,5 +89,34 @@ final class ContextStackTest extends PHPUnitTestCase
         self::assertSame($class, $context->peek());
         self::assertSame($project, $context->getProject());
         self::assertSame($typeContext, $context->getTypeContext());
+    }
+
+    /**
+     * @covers ::__construct
+     * @covers ::search
+     */
+    public function testSearchEmptyStackResultsInNull(): void
+    {
+        $project = new Project('myProject');
+        $context = new ContextStack($project);
+
+        self::assertNull($context->search(ClassElement::class));
+    }
+
+    /**
+     * @covers ::__construct
+     * @covers ::search
+     */
+    public function testSearchStackForExistingElementTypeWillReturnTheFirstHit(): void
+    {
+        $class = new ClassElement(new Fqsen('\MyClass'));
+        $project = new Project('myProject');
+        $context = new ContextStack($project);
+        $context = $context
+            ->push(new ClassElement(new Fqsen('\OtherClass')))
+            ->push($class)
+            ->push(new Method(new Fqsen('\MyClass::method()')));
+
+        self::assertSame($class, $context->search(ClassElement::class));
     }
 }

